### PR TITLE
Sample project using Hybrid Main Page

### PR DESCRIPTION
**Pull request by jbernoudy**
Updated SDK.Universal to include HybridMainPage.  Added sample project using HybridMainPage. Added VS2015 Solution.

### CODE REVIEW

**Comment by jbernoudy** [comment]
@smcnulty-sfdc @kchitalia

File: SalesforceSDK/Salesforce.SDK.Universal/Source/Hybrid/HybridMainPage.cs [hunk 1]
```diff
@@ -235,7 +240,13 @@ private void LoadUri(Uri uri)
             browser.NavigationStarting += onBrowserNavigationStarting;
             if (browser.Source == null)
             {
-                browser.Navigate(uri);
+                try
+                {
+                    browser.Navigate(uri);
+
+                }
+                catch (ArgumentException)
```
Comment (kchitalia) [inline SalesforceSDK/Salesforce.SDK.Universal/Source/Hybrid/HybridMainPage.cs]: Do we need to do anything in the catch block? What happens if browser.Navigate fails?

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config [hunk 2]
```diff
@@ -0,0 +1,23 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (kchitalia) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config]: Is this is a Windows10 specific file? If yes then lets make sure we dont add anything that is specific to Windows 10 to the solution. Users can add it manually via Add->Exisiting Project.

@smcnulty-sfdc What are your thoughts on this?

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/error.html [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿<!DOCTYPE html>
```
Comment (kchitalia) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/error.html]: Please add copyright info.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/index.html [hunk 4]
```diff
@@ -0,0 +1,25 @@
+﻿<!DOCTYPE html>
```
Comment (kchitalia) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/index.html]: Please add copyright info.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/default.js [hunk 5]
```diff
@@ -0,0 +1,214 @@
+﻿// For an introduction to the Blank template, see the following documentation:
```
Comment (kchitalia) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/default.js]: Please add copyright info.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
+var oauth = new SalesforceJS.OAuth2();
```
Comment (kchitalia) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js]: What is the difference between this file and default.js?

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/Settings/Config.cs [hunk 7]
```diff
@@ -26,39 +26,51 @@
...
-            string host = uri.Host;
-            int delimiter = host.LastIndexOf('_');
-            string encodedContentId = host.Substring(delimiter + 1);
-            IBuffer buffer = CryptographicBuffer.DecodeFromHexString(encodedContentId);
+            get { return "SfdcMobileWindowsRT"; }
+        }
+
+        public override string CallbackUrl
+        {
+            get { return "sfdc:///axm/detect/oauth/done"; }
```
Comment (bhariharan) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/Settings/Config.cs]: Why do we have ```Salesforce1``` specific ```callbackUrl``` here? This should be removed.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/Settings/Config.cs [hunk 8]
```diff
@@ -26,39 +26,51 @@
...
+    internal class Config : SalesforceConfig
     {
-        public IAsyncOperation<IInputStream> UriToStreamAsync(Uri uri)
+        public override string ClientId
         {
-            string host = uri.Host;
-            int delimiter = host.LastIndexOf('_');
-            string encodedContentId = host.Substring(delimiter + 1);
-            IBuffer buffer = CryptographicBuffer.DecodeFromHexString(encodedContentId);
+            get { return "SfdcMobileWindowsRT"; }
```
Comment (bhariharan) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/Settings/Config.cs]: Again - the ```clientId``` has to be configurable - not ```Salesforce1```.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/Settings/Config.cs (hunk 7 above)
Comment (bhariharan) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/Settings/Config.cs]: If a default value must be specified, we should use the values from ```bootconfig.json``` for the Mobile SDK sample apps.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    return function () { switchUser(account); }
+}
+
+var fetchContacts = function () {
+    var soql = 'SELECT Id, Name FROM Contact LIMIT 10';
+    var rest = Salesforce.SDK.Hybrid.Rest;
+    var cm = new rest.ClientManager();
+    var client = cm.peekRestClient();
+
+    var request = rest.RestRequest.getRequestForQuery("v31.0", soql);
```
Comment (bhariharan) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js]: Let's replace ```v31.0``` with a variable for API version if possible. It makes it easier to bump up the API versions with releases.

**Comment by jbernoudy** [comment]
I will have an updated pull request based on feedback tomorrow.  Thanks for the feedback.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config (hunk 2 above)
Comment (jbernoudy) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config]: @kchitalia This project is only included in the VS2015 solution so I am not sure why we would not include it.  There is still a separate VS2013 solution.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js (hunk 6 above)
Comment (jbernoudy) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js]: There isn't a default.js in this project.  I just started with a sample that used index, which seems appropriate for a non WinJS project.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config (hunk 2 above)
Comment (kchitalia) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config]: On my previous comment I mentioned that we should not include it if this is a Windows10 specific file. If its not then there is no harm in including it. 

So if you are using VS2015 on Windows 8.1, this solution will still work?

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js (hunk 6 above)
Comment (kchitalia) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js]: I see that there is a default.js file too. SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/default.js

Comment (jbernoudy) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/www/js/index.js]: That was meant for deletion.  Thanks for noticing it.

File: SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config (hunk 2 above)
Comment (jbernoudy) [inline SalesforceSDK/Salesforce.Sample.HybridMainPage.Universal/ApplicationInsights.config]: I didn't consider that scenario, I will test it once I get VS2015 installed on my windows 8.1 machine.

**Comment by jbernoudy** [comment]
Updated pull request with suggested fixes.  @bhariharan @kchitalia @smcnulty-sfdc

File: SalesforceSDK/Salesforce.SDK.Universal/Source/Hybrid/HybridMainPage.cs [hunk 10]
```diff
@@ -226,7 +226,12 @@ protected void LoadLocalStartPage()
             Uri url = browser.BuildLocalStreamUri(StreamResolverKey, _bootConfig.StartPage);
 
             // Pass the resolver object to the navigate call.
-            browser.NavigateToLocalStreamUri(url, new StreamUriResolver());
+            // browser.NavigateToLocalStreamUri(url, new StreamUriResolver());
```
Comment (bhariharan) [inline SalesforceSDK/Salesforce.SDK.Universal/Source/Hybrid/HybridMainPage.cs]: Let's remove this commented out ```LOC``` if it's not required.

**Comment by jbernoudy** [comment]
@bhariharan @kchitalia @smcnulty-sfdc. Updated pull request.

**Comment by kchitalia** [comment]
:+1:

**Comment by bhariharan** [comment]
Looks like you need to rebase and resubmit this pull request. It can't be merged anymore.

**Comment by jbernoudy** [comment]
@smcnulty-sfdc @kchitalia @bhariharan

**Comment by bhariharan** [comment]
:+1:

**Comment by kchitalia** [comment]
:+1:

**Pull request merged**